pkgs/applications/audio/quodlibet/default.nix

```diff
@@ -59,79 +63,113 @@ python3.pkgs.buildPythonApplication rec {
-    "--deselect=tests/test_browsers_iradio.py::TInternetRadio::test_click_add_station"
-    # upstream does actually not enforce source code linting
-    "--ignore=tests/quality"
-  ] ++ lib.optionals (withXineBackend || !withGstPlugins) [
-    "--ignore=tests/plugin/test_replaygain.py"
+  patches = [
+    (fetchpatch {
+      name = "python-3.12-startup.patch";
+      url = "https://patch-diff.githubusercontent.com/raw/quodlibet/quodlibet/pull/4358.patch";
+      hash = "sha256-3IjtAX2mKO/Xi/iTwT5WBD5CMTRYFED7XMm/cx+29Zc=";
+    })
+    (fetchpatch {
+      name = "more-python-3.12-fixes.patch";
+      url = "https://patch-diff.githubusercontent.com/raw/quodlibet/quodlibet/pull/4364.patch";
+      hash = "sha256-VRIQ+4e+X0kjZYuxV2wEjrFr+x5biwBtIR50K6hSfCY=";
+      excludes = [ "poetry.lock" ];
+    })
   ];
 
+  build-system = [ python3.pkgs.setuptools ];
+
+  nativeBuildInputs =
+    [
+      gettext
+      gobject-introspection
+      wrapGAppsHook3
+    ]
+    ++ (with python3.pkgs; [
+      sphinx-rtd-theme
+      sphinxHook
+    ]);
+
+  buildInputs =
+    [
+      adwaita-icon-theme
+      gdk-pixbuf
+      glib
+      glib-networking
+      gtk3
+      gtksourceview
+      kakasi
+      keybinder3
+      libappindicator-gtk3
+      libmodplug
+      libsoup
+      webkitgtk
+    ]
+    ++ lib.optionals (withXineBackend) [ xine-lib ]
+    ++ lib.optionals (withGstreamerBackend) (
+      with gst_all_1;
+      [
+        gst-plugins-base
+        gstreamer
+      ]
+      ++ lib.optionals (withGstPlugins) [
+        gst-libav
+        gst-plugins-bad
+        gst-plugins-good
+        gst-plugins-ugly
+      ]
+    );
+
+  dependencies =
+    with python3.pkgs;
+    [
+      feedparser
+      gst-python
+      mutagen
+      pycairo
+      pygobject3
+      setuptools
+    ]
+    ++ lib.optionals withDbusPython [ dbus-python ]
+    ++ lib.optionals withMusicBrainzNgs [ musicbrainzngs ]
+    ++ lib.optionals withPahoMqtt [ paho-mqtt ]
+    ++ lib.optionals withPypresence [ pypresence ]
+    ++ lib.optionals withSoco [ soco ];
+
+  nativeCheckInputs =
+    [
+      dbus
+      gdk-pixbuf
+      glibcLocales
+      hicolor-icon-theme
+      xvfb-run
+    ]
+    ++ (with python3.pkgs; [
+      polib
+      pytest
+      pytest-xdist
+    ]);
+
+  pytestFlags =
+    [
+      # missing translation strings in potfiles
+      "--deselect=tests/test_po.py::TPOTFILESIN::test_missing"
+      # require networking
+      "--deselect=tests/plugin/test_covers.py::test_live_cover_download"
+      "--deselect=tests/test_browsers_iradio.py::TInternetRadio::test_click_add_station"
+      # upstream does actually not enforce source code linting
+      "--ignore=tests/quality"
+    ]
+    ++ lib.optionals (python3.pythonAtLeast "3.12") [
+      # gdist module is broken with python 3.12
+      "--ignore=tests/test_po.py"
+      "--ignore=tests/test_setup.py"
+      "--deselect=tests/plugin/test_covers.py::TCovers::test_cover_path_lastfm"
+    ]
```

Comment:
```suggestion
```

Here's a patch that imports `setuptools` before `distutils` in `gdist/__init__.py`:

```
From b3980cf4c8766815e4c13ee1695cebfbf473a2e6 Mon Sep 17 00:00:00 2001
From: Pavel Sobolev <contact@paveloom.dev>
Date: Sun, 14 Jul 2024 11:30:11 +0300
Subject: [PATCH] Import `setuptools` before importing `distutils`.

---
 gdist/__init__.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/gdist/__init__.py b/gdist/__init__.py
index 71814af49..a54aac653 100644
--- a/gdist/__init__.py
+++ b/gdist/__init__.py
@@ -31,6 +31,7 @@ Also supports setuptools but needs to be imported after setuptools
 (which does some monkey patching)
 """

+import setuptools
 import sys

 from distutils.core import setup
--
2.45.2

```